README.md
dialog
======

Fork of Ruby gem dialog


ChangeLog
2008-03-29  Martin Landers  <elk@treibgut.net> (Release 0.1.2)

	* Rakefile: Bumped revision to 0.1.2.
	* test/test.rb: Added an example that uses MixedForm.

2008-03-29  R.T. Lechow <podoboo@gmail.com>

	* lib/mixedform.rb: New file, adds --mixedform support.
	* lib/dialog.rb: Added mixedform.rb to includes.

2008-03-24  Martin Landers  <elk@treibgut.net>

	* ChangeLog: New file.
	* form.rb: Fixed example in comment.

2008-03-03  Martin Landers  <elk@treibgut.net> (Release 0.1.1)

	* Rakefile: Updated E-Mail address.
	Bumped revision to 0.1.1.
	* lib/base.rb: Updated E-Mail address.
	* README: Updated E-Mail address.
	Updated copyright clause.

2006-10-20  Martin Landers  <elk@treibgut.net> (Release 0.1.0)

	* lib/*.rb: Initial implementation of the gem.
	* test/*.rb: Initial "tests", only manual for now.
	* COPYING: New file, GPL license.
	* Rakefile: Initial buildfile for the gem.
	* README: RDoc for the gem.


test/test.rb
require 'rubygems'
require 'dialog'

include Dialog

# Test MixedForm
steps = MixedForm.new do |m|
  m.text "Authentication"
  m.field "User", 1, 1, "", 1, 10, 40, 38, 0
  m.password "Password", 2, 1, "", 2, 10, 40, 38
end
steps.show!
 
# Test menu
steps = Menu.new(0,70) do |m|
  # m.backtitle "Bootable Backup"
  # m.title "Configure steps"
  m.text "Bla bla"
  m.item_help
  m.defaultno
  m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage", "Help text"
  m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd", "More help text"
  m.ok { |d,m| puts "Hey, you pressed ok and selected #{m}" }
  m.cancel { |d,m| puts "Oh no! You canceled this wonderful dialog! Choice was #{m}" }
  m.extra("Maybe") { |d,m| puts "Hm, now that's extra-ordinary. Selected #{m}" }
  m.help { |d,m| puts "Help! I need somebody, Help! For #{m}" }
end
steps.show!

# Test yesno
yn = Yesno.new do |t|
  t.text "Do you like the new ruby dialog gem?"
  t.ok { puts "Hey, you pressed ok" }
  t.cancel { |d| puts "Oh no! You canceled this wonderful dialog!" }
  t.extra { puts "Hm, now that's extra-ordinary" }
  t.help { puts "Help! I need somebody, Help!" }
end
yn.show!

# Test timebox
tb = Timebox.new(0, 0) do |t|
  t.text "What time is it?"
  t.time Time.now + 5*60
end
tb.show!

# Test textbox
aFile = File.new("foo.txt", "w")
99.downto(1) do |v|
  aFile.puts "#{v} bottles of beer on the wall, #{v} bottles of beer, ..."
end
aFile.flush
aFile.close

textbox = Textbox.new(30, 50) do |t|
  t.text "Please read the following aloud"
  t.file aFile
end
textbox.show!

File.unlink(aFile.path)


# Test tailboxbg
aFile = File.new("foo.txt", "w")

tailbox = Tailboxbg.new(30, 60) do |t|
  t.text "And now for some rubbish from our parent..."
  t.file aFile
end
tailbox.show

99.downto(1) do |v|
  aFile.puts "#{v} bottles of beer on the wall, #{v} bottles of beer, ..."
  sleep 0.01
  aFile.flush
end

tailbox.wait
aFile.close
File.unlink(aFile.path)


# Test tailbox
aFile = File.new("foo.txt", "w")

tailbox = Tailbox.new(30, 60) do |t|
  t.text "And now for some rubbish from our parent..."
  t.file aFile
end
tailbox.show

99.downto(1) do |v|
  aFile.puts "#{v} bottles of beer on the wall, #{v} bottles of beer, ..."
  sleep 0.01
  aFile.flush
end

tailbox.wait
aFile.close
File.unlink(aFile.path)


# Test radiolist
rl = Radiolist.new do |m|
  m.text "Speak friend and enter"
  m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage", :on
  m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd"
  m.choice "/dev/hda3", "Backup partition /dev/hda3 using dd", :off
end
rl.show!

# Test passwordbox
pwbox = Passwordbox.new do |b|
  b.text "Speak friend and enter"
end
pwbox.show!

# Test msgbox
box = Msgbox.new(50,50) do |m|
  m.text "Sir, a message for you!"
end
box.show!

# Test menu
steps = Menu.new do |m|
  # m.backtitle "Bootable Backup"
  # m.title "Configure steps"
  m.text "Bla bla"
  m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage"
  m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd" 
end
steps.show!

# Test inputmenu
steps = Inputmenu.new(0, 75) do |m|
  m.text "Bla bla bla"
  m.menu_height 20
  m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage"
  m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd" 
end
steps.show!

# Test inputbox
ibox = Inputbox.new do |b|
  b.text "Your input please"
  b.value "Default input"
end
ibox.show!

# Test gauge
gauge = Gauge.new do |g|
  g.text "Please wait... magic in progress"
  g.complete 0.1
end
gauge.show

10.step(100, 10) do |v|
  gauge.complete v
  sleep 0.05
end
gauge.wait

# Test file selector
sel = FSelect.new do |f|
  f.text "Choose a file"
  f.file "./bla.txt"
end
sel.show!

# Test Form
frm = Form.new do |f|
  f.text "Bla Bla"
  f.field "Fruit", 1, 1, "Strawberry", 1, 12
  f.field "Veggie", 2, 1, "Eggplant", 2, 12, 30, 50
end
frm.show!

# Test Checklist
chklist = Checklist.new do |m|
  # m.backtitle "Bootable Backup"
  # m.title "Configure steps"
  m.text "Bla bla"
  m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage", :on
  m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd"
  m.choice "/dev/hda3", "Backup partition /dev/hda3 using dd", :off
end
chklist.show!

# Test calendar
cal = Calendar.new
cal.date Date.today - 5
cal.show!

# Test infobox
ibox = Infobox.new do |b|
  b.text "That's all folks!"
end
ibox.show!


test/monthy.rb
require 'rubygems'
require 'dialog'

include Dialog

# Everybody's favourite Monty Python question in Yes/No style
yn = Yesno.new do |yn|
  yn.text "What is your favourite colour?"
  yn.yes "Blue"
  yn.no "Yellow"
end
yn.show!

if yn.yes?
  puts "Go on. Off you go."
else
  puts "auuuuuuuugh."
  exit
end


# As a menu
menu = Menu.new do |m|
  m.text "What... is your favourite colour?"
  m.choice "blue", "Blue"
  m.choice "yellow", "Yellow"
end
menu.show!

if menu.yes? && menu.output == "blue"
  puts "Go on. Off you go."
else
  puts "auuuuuuuugh."
  exit
end


# And as a menu using blocks as handlers
menu = Menu.new do |m|
  m.text "What... is your favourite colour?"
  m.choice "blue", "Blue"
  m.choice "yellow", "Yellow"
  m.no { |d,o| puts "auuuuuuuugh." }
  m.yes do |d,o|
    if o == "blue"
      puts "Go on. Off you go."
    else
      puts "auuuuuuuugh."
      exit
    end
  end
end
menu.show!


test/defaults.rb
require 'dialog'

include Dialog

Dialog::default_options do |d|
  d.backtitle "Camelot!"
end

Yesno.new do |yn|
  yn.text "What is your favourite colour?"
  yn.yes "Blue"
  yn.no "Yellow"
end.show!


template = DialogOptions.new do |d|
  d.backtitle "Camelot Template!"
end

Yesno.new(0,0,template) do |yn|
  yn.text "What is your favourite colour?"
  yn.yes "Blue"
  yn.no "Yellow"
end.show!


lib/menu.rb
require 'base'

module Dialog

  # Displays a menu.
  #
  # Corresponds to the --menu box option. Use the choice method to
  # create menu items.
  #
  # Example:
  #
  #   steps = Dialog::menu do |m|
  #     m.text "Bla bla"
  #     m.choice "/dev/hda1", "Backup partition /dev/hda1 using partimage"
  #     m.choice "/dev/hda2", "Backup partition /dev/hda2 using dd" 
  #   end
  #
  # Box option syntax:
  #   --menu <text> <height> <width> <menu height> <tag1> <item1>...
  #
  class Menu < Base

    def initialize(*options)
      @choices = []
      super(*options)
      @options.box_options[3] ||= 0 # Default menu-height 
    end

    # Adds a choice to the menu
    def choice(tag, item, help=nil)
      choice = [tag, item]
      choice << help unless help.nil?
      @choices << choice
    end

    def box_options
      super
      @choices.flatten
    end

    def menu_height(v)
     @options.box_options[3] = v 
    end

  end

end


lib/util.rb
module Dialog
  module Util

    # A more flexible variant of the popen3 call
    #
    # Other than Open3.popen3, this method allows the parent process
    # to share individual file descriptors with the child process. By
    # default, all three file descriptors (stdin, stdout, stderr) of
    # the child process are connected to pipes, which the parent process
    # can use to write to/read from the child process. However, if this
    # is not the desired behaviour, the parent process can pass IO objects
    # for each stream that should be shared by the child process. This
    # method does not fork twice (putting the burden to wait/waitpid on
    # the parent process) and thus gives the parent process a pid. Note,
    # that this introduces a small chance of receiving yet unitinialized
    # file descriptors.
    #
    # Note, that popen3 behaves like exec concerning its arguments. If popen3
    # is given a single argument, that argument is taken as a line that is subject
    # to shell expansion before being executed. If multiple arguments are given,
    # the second and subsequent arguments are passed as parameters to the command
    # without shell expansion taking place.
    #
    # Examples:
    #   
    #   # Share stdin/stdout, but connect stderr to a pipe
    #   # The * is subject to shell file globbing
    #   pid, stdin, stdout, stderr = popen3("echo *", :stdin => STDIN, :stdout => STDOUT)
    #
    #   # Share stdin/stdout, but connect stderr to a pipe
    #   # Echoes an asterisk
    #   pid, stdin, stdout, stderr = popen3("echo", "*", :stdin => STDIN, :stdout => STDOUT)
    #
    def self.popen3(*command)
      fds = command.pop if command.last.kind_of?(Hash)
      command.flatten!
      stdin = fds[:stdin] if fds
      stdout = fds[:stdout] if fds
      stderr = fds[:stderr] if fds

      # pipe[0] for read, pipe[1] for write
      pi = IO::pipe unless stdin
      po = IO::pipe unless stdout
      pe = IO::pipe unless stderr

      pid = fork

      if pid.nil?
	# child
        pi[1].close if pi
        STDIN.reopen(pi[0]) if pi
        pi[0].close if pi

        po[0].close if po
        STDOUT.reopen(po[1]) if po
        pi[1].close if po

        pe[0].close if pe
        STDERR.reopen(pe[1]) if pe
        pe[1].close if pe

	exec(*command)
      end

      # parent
      pi[0].close if pi
      po[1].close if po
      pe[1].close if pe
      pi[1].sync = true if pi
      return [pid, stdin || pi[1], stdout || po[0], stderr || pe[0]]
    end

  end
end


lib/gauge.rb
require 'base'

module Dialog

  # Displays a gauge.
  #
  # Corresponds to the --gauge box option. Use the complete method
  # to set the initially completed part of the gauge. TODO How do
  # we update the gauge??
  #
  # Example:
  #
  #   steps = Dialog::gauge do |m|
  #     m.text "Please wait"
  #     m.complete 30  # 30% complete, 0.3 works as well
  #   end
  #
  # Box option syntax:
  #   --gauge <text> <height> <width> [<percent>]
  #
  class Gauge < Base

    # Sets the value to display
    def complete(p)
      @completed = case p
        when Fixnum then p
        when Bignum then p
        when Float then (p*100).to_i
        else p
      end
      @stdin.puts @completed if @stdin 
    end

    def arguments
      super @completed || 0
    end

    # Make sure wait returns by closing stdin
    def wait
      @stdin.close if @stdin
      super
    end

  end

end


lib/form.rb
require 'base'

module Dialog

  # Displays a form
  #
  # Corresponds to the --form box option. Use the field method to
  # define form fields
  #
  # Example
  #
  #   frm = Form.new do |f|
  #     f.text "Pick your poi.., er, vitamins"
  #     f.field "Fruit", 1, 1, "Strawberry", 1, 12
  #     f.field "Veggie", 2, 1, "Eggplant", 2, 12, 30, 50
  #   end
  #
  # Box option syntax:
  #   --form <text> <height> <width> <form height> <label1> <l_y1> <l_x1> <item1> <i_y1> <i_x1> <flen1> <ilen1>...
  #
  class Form < Base

    def initialize(*options)
      @fields = []
      super(*options)
      @options.box_options[3] ||= 0 # Default form-height
    end

    # Adds a field to the form
    def field(label, ly, lx, item, iy, ix, flen=30, ilen=0)
      @fields << [label, ly, lx, item, iy, ix, flen, ilen]
    end

    def box_options
      super
      @fields.flatten
    end

    def form_height(v)
      @options.box_options[3] = v
    end

  end

end
